compute/src/main/java/org/jclouds/compute/domain/internal/ArbitraryCpuRamTemplateBuilderImpl.java

```diff
@@ -0,0 +1,82 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.compute.domain.internal;
+
+import com.google.common.base.Supplier;
+import org.jclouds.collect.Memoized;
+import org.jclouds.compute.domain.Hardware;
+import org.jclouds.compute.domain.HardwareBuilder;
+import org.jclouds.compute.domain.Image;
+import org.jclouds.compute.domain.Processor;
+import org.jclouds.compute.domain.TemplateBuilder;
+import org.jclouds.compute.options.TemplateOptions;
+import org.jclouds.compute.util.AutomaticHardwareIdSpec;
+import org.jclouds.domain.Location;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import java.util.Set;
+
+import static org.jclouds.compute.util.AutomaticHardwareIdSpec.generateAutomaticId;
+import static org.jclouds.compute.util.AutomaticHardwareIdSpec.isAutomaticId;
+import static org.jclouds.compute.util.AutomaticHardwareIdSpec.parseId;
+
+public class ArbitraryCpuRamTemplateBuilderImpl extends TemplateBuilderImpl {
+   @Inject
+   protected ArbitraryCpuRamTemplateBuilderImpl(@Memoized Supplier<Set<? extends Location>> locations,
+                                                @Memoized Supplier<Set<? extends Image>> images, @Memoized Supplier<Set<? extends Hardware>> hardwares,
+                                                Supplier<Location> defaultLocation, @Named("DEFAULT") Provider<TemplateOptions> optionsProvider,
+                                                @Named("DEFAULT") Provider<TemplateBuilder> defaultTemplateProvider) {
+
+      super(locations, images, hardwares, defaultLocation, optionsProvider, defaultTemplateProvider);
+   }
+
+   protected Hardware automaticHardwareForCpuAndRam(double cores, int ram) {
+      return new HardwareBuilder()
+            .id(generateAutomaticId(cores, ram))
+            .ram(ram)
+            .processor(new Processor(cores, 1.0))
+            .build();
+   }
+
+   protected Hardware findHardwareWithId(Set<? extends Hardware> hardwaresToSearch) {
+      try {
+            return super.findHardwareWithId(hardwaresToSearch);
+      } catch (Exception NoSuchElementException) {
+            if (isAutomaticId(hardwareId)) {
+               AutomaticHardwareIdSpec spec = parseId(hardwareId);
+               return automaticHardwareForCpuAndRam(spec.getCores(), spec.getRam());
+            }
+            else{
+               throw throwNoSuchElementExceptionAfterLoggingHardwareIds("Invalid hardwareId", hardwaresToSearch);
+            }
+        }
+    }
+
+   protected Hardware resolveHardware(Set<? extends Hardware> hardwarel, final Iterable<? extends Image> images) {
+      try {
+         return super.resolveHardware(hardwarel, images);
+      }
+      catch (Exception NoSuchElementException) {
+         if (super.minCores != 0 && super.minRam != 0) {
```

Comment:
Considering that if only minRam or minCores are provided and they match with a hardware profile a VM is created according to the hardware profile, I think that we should have some defaults. I will ask on the mail list about the default values.